### PR TITLE
Add hunger

### PR DESCRIPTION
**Pull request by BlockMen**
Some hunger implementation i wrote a while ago.
The current state is WIP (more focused on way of implementation) hunger system needs improvement.

Don't remember the state i left the code back then, tested only building and player serialization today.

### CODE REVIEW

**Comment by MoNTE48** [comment]
Finally, it happened! I'll test it later. Question: where the file bread.png?

File: builtin/game/hunger.lua [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+-- GNU Lesser General Public License for more details.
+--
+-- You should have received a copy of the GNU Lesser General Public License along
+-- with this program; if not, write to the Free Software Foundation, Inc.,
+-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+---------------------------------------------------------------------------------
+
+
+-- Dont do anything when no damage or hunger
+if core.setting_getbool("enable_damage") ~= true then
```
Comment (RobertZenz) [inline builtin/game/hunger.lua]: Did you forget to add a check for `enable_hunger` or something similar here?

**Comment by Calinou** [comment]
:+1:. Perhaps it could be toggle-able from GUI for singleplayer games?

File: doc/lua_api.txt [hunk 2]
```diff
@@ -1892,7 +1892,7 @@ Call these functions only at load time!
       * The provided function should check that the position is protected by the mod
         calling this function before it prints a message, if it does, to allow for
         multiple protection mods.
-* `minetest.register_on_item_eat(func(hp_change, replace_with_item, itemstack, user, pointed_thing))`
+* `minetest.register_on_item_eat(func(hunger_change, replace_with_item, itemstack, user, pointed_thing, hp_change))`
```
Comment (est31) [inline doc/lua_api.txt]: You should have it in different order. This way the API breaks.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -2096,7 +2096,7 @@ and `minetest.auth_reload` call the authetification handler.
 * `minetest.create_detached_inventory(name, callbacks)`: returns an `InvRef`
     * callbacks: See "Detached inventory callbacks"
     * Creates a detached inventory. If it already exists, it is cleared.
-* `minetest.do_item_eat(hp_change, replace_with_item, itemstack, user, pointed_thing)`:
+* `minetest.do_item_eat(hunger_change, replace_with_item, itemstack, user, pointed_thing, hp_change, eat_sound)`:
```
Comment (est31) [inline doc/lua_api.txt]: Here too.

File: doc/lua_api.txt [hunk 4]
```diff
@@ -2211,8 +2211,8 @@ These functions return the leftover itemstack.
     * returns `itemstack, success`
 * `minetest.item_drop(itemstack, dropper, pos)`
     * Drop the item
-* `minetest.item_eat(hp_change, replace_with_item)`
-    * Eat the item.
+* `minetest.item_eat(hunger_change, replace_with_item, hp_change, eat_sound)`
```
Comment (est31) [inline doc/lua_api.txt]: and here.

File: src/script/lua_api/l_object.cpp [hunk 5]
```diff
@@ -1051,6 +1052,41 @@ int ObjectRef::l_get_breath(lua_State *L)
 	return 1;
 }
 
+// set_hunger(self, hunger)
+int ObjectRef::l_set_hunger(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	PlayerSAO* co = getplayersao(ref);
```
Comment (est31) [inline src/script/lua_api/l_object.cpp]: Style issue: there is a space needed before the star.

File: src/script/lua_api/l_object.cpp [hunk 6]
```diff
@@ -1051,6 +1052,41 @@ int ObjectRef::l_get_breath(lua_State *L)
...
+
+	return 0;
+}
+
+// get_hunger(self)
+int ObjectRef::l_get_hunger(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	PlayerSAO* co = getplayersao(ref);
```
Comment (est31) [inline src/script/lua_api/l_object.cpp]: same here

File: builtin/game/item.lua [hunk 7]
```diff
@@ -1,4 +1,22 @@
--- Minetest: builtin/item.lua
+---------------------------------------------------------------------------------
```
Comment (est31) [inline builtin/game/item.lua]: None of the files in builtin/ have a copyright header. And if you add one, can you at least watch out that the year matches the current one?

File: doc/lua_api.txt (hunk 2 above)
Comment (BlockMen) [inline doc/lua_api.txt]: @est31 it does not break the API, see https://github.com/minetest/minetest/pull/3388/files#diff-83bb4564c6fbf230569bdeb2becc32b9R435

When hunger is disabled it falls back to old behavior. You can't expect to add a hunger system and keep the healing by default then. A fallback is enough to not break anything.

File: builtin/game/item.lua (hunk 7 above)
Comment (BlockMen) [inline builtin/game/item.lua]: That not right. Only builtin/game/ has no headers yet, everywhere else are often some. In mainmenu 99% have.
I would have said that it is the state of 2013, but ok, i can change it to 2015. Also moving it out of this PR might be good.

File: builtin/game/hunger.lua (hunk 1 above)
Comment (BlockMen) [inline builtin/game/hunger.lua]: Yep :)

File: doc/lua_api.txt (hunk 2 above)
Comment (est31) [inline doc/lua_api.txt]: So its different depending on the setting? That must be documented.

Comment (BlockMen) [inline doc/lua_api.txt]: https://github.com/minetest/minetest/pull/3388/files#diff-1305560bd8befb32862f0feeefabd02eR2215

But i agree it should be made more clear ^

**Comment by PilzAdam** [comment]
:-1:, I would prefer a generic way to attach arbitrary attributes to players.

File: builtin/game/item.lua (hunk 7 above)
Comment (est31) [inline builtin/game/item.lua]: @BlockMen  lets discuss it in #3390 . And if we change, lets change it for the whole builtin/ directory.

**Comment by BlockMen** [comment]
@PilzAdam why didn't you implement it when you were adding drowning?

**Comment by est31** [comment]
On the long term, I think @PilzAdam is right. We should remove all values (hp, hunger, breath) from core, and make them definable by mods. On the short to mid term, idk.

**Comment by 0-afflatus** [comment]
Long term, what @PilzAdam and @est31 said.
Short term, I would rather use a simple builtin than the hunger mod.
If it is kept simple, means not having to redefine on_eat and is optional (ideally with a menu tickbox next to creative mode and enable damage), then I'd welcome this.

Hunger is such a core mechanism that it should be directly supported by the engine and be available in the default game. If it can be implemented with a view to future generalisation, all the better. As a sub-game designer it would save me the trouble of integrating the hunger mod.

**Comment by ShadowNinja** [comment]
Can this be implemented fairly nicely as a separate mod or is core support really necessary?

**Comment by BlockMen** [comment]
Define "fairly nicely". Ofc it can be done as mod (see http://github.com/BlockMen/hud_hunger), but its not really clean doable. Hunger is a often requested thing and should be implemented (at least in a basic way) by the engine, to allow easy control and handling by users.

Main problems by current mod implementations are the need to override many things, starting by the items moving over HUD statbars and ending in abusing inventory or bloated stat files.

IMO hunger is a core feature (since we have somewhat an implementation of survival aspects). I chose the way that allows flexibility and relieablity, so mods could override the hunger function (at least at current state).

**Comment by 0-afflatus** [comment]
Hunger should be implemented on the same level as damage and drowning. Doing it as a bolt-on extra (i.e. a mod) is sub-optimal. Is there any reason core support would not be desirable?

**Comment by PilzAdam** [comment]
I just looked at this a bit closer. There is nothing that this does better than any mod implementation. The whole logic runs on the server; there is nothing gained from moving this from mods to the engine.

If there was something client-side involved int this, then I would at least understand why people want this to be moved to the core. But there isn't.

@ShadowNinja core support does not give any advantage here, see above.

@BlockMen What is bad about "bloated stat files"? It's perfectly valid and wanted that mods can write their serialized data to text files in the world directory.
Also you are moving these problems just to another place, from mods to the engine.

@0-afflatus drowning needed to be done in th engine because it *required* client-side stuff, which is currently not possible with mods. Hunger, on the other hand, does not, so these two are not equal in this aspect.

**Comment by BlockMen** [comment]
@PilzAdam ofc drowning was possible with mods, but it was slow. And the current implementation can be easily hacked so that players can disable drowning completely on every server. So having everything clientside is not "the solution" aswell.

Also, you completly ignore the advantages i claimed of moving into core (except the player stat files), nice. And concerning them, one stat more in the player file is smaller than more files for each different mod.

PS: are you sure that you opinion on hunger at all is not the reason to argue this way?

**Comment by 0-afflatus** [comment]
@PilzAdam ok, I get that core support for hunger is not absolutely necessary and that it is not a feature that you would use (reading between the lines here). I still don't understand why it would be a problem.

As I'm not a collaborator here I don't need a re-run of previous discussions, but some pointers to them would be handy.

My concerns would be around the details of this implementation, which seem rather complex, and its modifiability.

**Comment by rubenwardy** [comment]
These sorts of things should be controlled be subgames and mods, not builtin, imo.

**Comment by 4Evergreen4** [comment]
I agree with @rubewardy.  The engine currently has way too much stuff that is specifically for a minecraft style of gameplay.  For instance, what if someone wanted to make a RTS game in minetest and didn't want hunger or drowning?  Even if you give the option to disable it, it still should be something that is accomplished by using the lua API.

**Comment by BlockMen** [comment]
@rubenwardy, and the reason(s) for your opinions is/are?

@4Evergreen4, MC like games are the main usage of the engine, so adding features for this usecase
seem natural. Same goes for other game types, but since there are no specific requests (for features) and those "MC features" can be disabled completely by Lua i see no point in not adding.


@all, could someone of the "not to core" fraction tell me a reason in which way this addition will
harm the engine or builtin?

@celeron55 your opinion on this? In case you are against adding hunger support to engine aswell we can finish this and i will close the PR.

**Comment by PilzAdam** [comment]
Side note: Adding a generic system for stuff like this isn't really that hard; the time and effort you spend on convincing people to merge this would be enough to just implement the generic system.

**Comment by BlockMen** [comment]
@PilzAdam sidenote2: Why don't you answer my questions?

**Comment by rubenwardy** [comment]
There is no benefit at all to using this over a mod. It's just more code in builtin that shouldn't be there. If you need to add a feature like this, it's a sign that your API isn't good enough and needs improving.

**Comment by rubenwardy** [comment]
> the need to override many things

minetest.register_on_item_eat was designed for hunger. You could also override minetest.do_item_eat.

> starting by the items moving over HUD statbars

This is a sign that HUD sucks.

> and ending in abusing inventory or bloated stat files.

This is a sign that mod data storage sucks.

**Comment by BlockMen** [comment]
> There is no benefit at all to using this over a mod. It's just more code in builtin that shouldn't be there. If you need to add a feature like this, it's a sign that your API isn't good enough and needs improving.

Wrong. Just the fact that it can be done by mods is no valid reason to not do it in core. Hunger is a core part in aspects of survival and in aspect of handling a basic thing as hunger system. Another example for those baisc core feature is ```core.is_protected()```.
This is also "useless" since it does nothing itself, but is added to have a common way of handling.

> This is a sign that HUD sucks.
> This is a sign that mod data storage sucks.

I could add something more: This is a sign that MT sucks.

Those (stupid) sentences just say: Let's stop adding features until the engine is perfect.

Deal?

**Comment by kilbith** [comment]
The hunger idea is fine, as a primary state.
Why not abstracting the drowning code and design an API for the player health attributes ?

Some draft : 

``` Lua
-- Called within an on_step-like callback
-- Drowning
{
  statbar_id = 2,  -- determine its position
  degrade = { 
          type = "active",
          active = {
                 player.inside = { "group:liquid" },
                 player.trigger_y = 1.4   -- head level
          }
  },
  recover = { type = "passive" },  -- air
  loss_step = 2,   -- one bubble loss by step
  loss_freq = 1.0,  -- step interval in seconds
  regain_step = 2,
  regain_freq = 0.2,
  textures = { "bubble_on.png", "bubble_off.png" }   -- omit "_off.png" leaves a blank space on degrading
}


-- Hunger
{
  statbar_id = 3,
  degrade = { type = "passive" },
  loss_step = 1,
  loss_freq = 30,
  textures = { "bread_on.png", "bread_off.png" }
}
```

**Comment by Ferk** [comment]
> This is a sign that mod data storage sucks.

Imho, this would help: #3374 

> I could add something more: This is a sign that MT sucks.

I think the idea is to try and fix it so it sucks less. The thing is that the problems you find for hunger would also be problematic for other mechanisms a custom game might want to add.

While I agree that hunger is a central feature for any survival experience, I also think that as an engine Minetest is suited for more than just survival games, what if someone wants to make a racing game and add status bars for gas and speed? or status bar for experience points in a rpg-like game?

I'd argue it might be better if even things like breathing and health could be moved to mods/games and have more powerful APIs with more client-side control, rather than bloating the engine with gameplay-specific logic.

As a side note: to be frank, I find it's rare to encounter a piece of software that doesn't suck in one way or another. Most of the time if there's software that you think it's good it's only because you didn't have a look at its code.

**Comment by 0-afflatus** [comment]
Having read the arguments so far I'm going with :-1: 

@BlockMen I get your frustration - I see how much work you have put into not only your hunger system, but the whole of minetest itself. With all due respect, (I mean that) I think you could do with spending some time in a non-procedurally generated natural environment.

Let's crack on (as soon as this forthcoming release is out of the door) with developing a generalised statistics API **with hunger as a primary use case**.

**Comment by BlockMen** [comment]
Made a forum topic to include the users more: https://forum.minetest.net/viewtopic.php?f=5&t=13644

@0-afflatus I'm not that much, it was kinda expectable. Everything long wanted is blocked, see 3D torches. The "it can be done as mod" makes Minetest for many users just one thing: Crap.

**Comment by rubenwardy** [comment]
If you want hunger in Minetest, then it should be in Minetest Game not
built into the engine.

**Comment by CasimirKaPazi** [comment]
I already wrote my general criticism in the forum (short: what PilzAdam said). Being more precise regarding the code:
Others already showed how it could work as a general status/health management. Those are the features you added in "Step 1". The "simple hunger mechanics" is the part that should be done by mod. The commit "Support for eating sound" could be a separate pull request, it is not necessary for the function of hunger.

**Comment by sapier** [comment]
Imho there should be mechanism for generic player properties not "add hunger" "add breath" "add money" ... for what I remember I did once implement something like that ... for sure it can't be used as is but it could replace health as well as breath and enable mods to implement whatever they want.

**Comment by sofar** [comment]
This pull request comes from a repository that is clearly marked as GPLv3, making it incompatible with the merge requirements for minetest code, which needs to be LGPL2.1+ for code (or, compatible with that).

As such, it cannot be merged as is.

**Comment by est31** [comment]
These two repos are different:

* https://github.com/blockplanet/blockplanet
* https://github.com/BlockMen/minetest

**Comment by sofar** [comment]
Happy to retract, that repo does not include a toplevel gplv3 statement.

Yes, I have no problems admitting I made a mistake.

**Comment by Fixer-007** [comment]
Add thirst :) If you need food, you need water also :}

**Comment by neoascetic** [comment]
If this going to be added, then mobs should be added too, since it is the most requested feature. Personally, I am :-1: on this. What is minetest? Is this a game, engine or what?

**Comment by C1ffisme** [comment]
> then mobs should be added too

We should add an API for things like mobs, and this.

Why not make values for the player, that can be added or disabled, like ```player.health```, etc.

Then we could do:

```
minetest.register_on_joinplayer(function(player)  -- This is pseudo code, because I haven't modded in a while.
    player.health = 20
    player.hunger = 20
    etc...
end)
```

**Comment by Dragonop** [comment]
:-1: This could be good if minetest_game had any other source of food aside from apples and farming (like mobs), but it doesn't, making it unbalanced.

**Comment by nerzhul** [comment]
:-1: we should close this i think

**Comment by BlockMen** [comment]
We have no rule to close when 2 core devs give :-1: . Reopening.

**Comment by Dragonop** [comment]
What are the rules for closing a pull request?

**Comment by tchncs** [comment]
For me, from outside-view and fully neutral, ```creative``` is part of the ```minetest_game``.
Following this logic, ```hunger``` can also be a part of it... .

**Comment by raymoo** [comment]
I'd be fine with this either in engine or in minetest_game, I just want there to be some canonical hunger mechanic.

**Comment by C1ffisme** [comment]
@Dragonop About your food comment, let's not use food (a Minetest Game flaw) as an excuse not to merge hunger into Minetest (the engine), rather, let's use this pull request to fix the Minetest Game flaw.

Overall, I support hunger, since the current gameplay gets boring in a couple minutes without any threat.

**Comment by sofar** [comment]
https://github.com/minetest/minetest_game/pull/855

**Comment by BlockMen** [comment]
https://github.com/minetest/minetest_game/pull/855#issuecomment-184592671

**Comment by MoNTE48** [comment]
You could also create a small mod to support 3d armour bar? And to make the substrate under health and hunger, as in the HUD?

**Comment by sofar** [comment]
I think this implementation isn't too bad, at all.

However, it completely doesn't solve the raised point of not being able to store player metadata, and so the next person to come along that wants to add some persistent player attribute will have to hack around the same problem again.

I'd love it if this patch could build upon a generic ```meta``` class. One that is the parent of ```NodeMeta```, ```ItemStackMeta``` and then also ```PlayerMeta```. The first one already exists, but is a single-trick pony metadata storage. The second one is one I envisioned to replace ItemStack:get_metadata(). Then we can serialize PlayerMeta into the worldfolder/players/playername file, and every mod can use and manipulate player stats and other player attributes to their hearts' content.

**Comment by 0-afflatus** [comment]
@sofar something like a generic PlayerMeta is the WTG as far as engine changes go.
I'm now feeling strongly (FWIW) that hunger should be included in minetest_game as an option and I'm coming round to Blockmen's suggested HUD improvements, but whatever the case, I am going to want to customise, particularly the HUD in grailtest, so these changes should not be in the engine, just the generic mechanisms to enable them.

**Comment by hunter688** [comment]
When player is human, need hungerbar and thirstbar. When someone create subgame - player as robot, will be better for example battery level, oil level etc. I think some generic properties of players.
When modder would like add for example radiation level, angry level or something, I prefer api for adding whatever modder want.

**Comment by neoascetic** [comment]
Great example. Why do we think player is always a human?

**Comment by Zeno-** [comment]
I think this should probably just be closed. It needs to be rebased, it's not finished, it contains bugs, and the comments (referring to the source code) have not been addressed and most likely won't be.

**Comment by Zeno-** [comment]
Unfinished work that needs issues resolved. The author of this work is very unlikely to do the work needed so I'm closing this.

If anyone wants to continue the work then the best approach is to apply the commit to their repo and continue the work.

**Pull request closed without merging**